### DISCUSSION TITLE
OS_API_Init() Failure when attempting to run cross-compiled code on an embedded device

### DISCUSSION THREAD

**Discussion by mc-bane** [nasa/cFS#336] [Q&A] [answered]
### Context

I've been attempting to run cFS on an Arduino Yun Rev 2's microprocessor (the yun has a normal microcontroller avr chip and an integrated serial connection to an atheros microprocessor running a linux-based OS OpenWrt I'd like cFS to live on). After successful cross-compiling I've run into an issue I've not run into when I run on my native linux host machine when I run the core-cpu1 executable on the Yun. 

Pic related is the output when I run ./core-cpu1 from the yun I am ssh'd into.
![Screenshot from 2021-08-11 15-40-58](https://user-images.githubusercontent.com/86273861/129092819-8d79b84a-c8bf-4acf-aa06-12a57bbbb55d.png)

### Question
Is there an OS setting from the sample_defs I need to change to not trigger the pthread_mutexattr_setprotocol error? I'm not really sure what is throwing this error.

Thanks!

**Comment by skliper** [comment, marked as answer]
Error is from this section of code:
https://github.com/nasa/osal/blob/ce11e7a20d4fd9176469659acb1ee3b371ba4291/src/os/posix/src/os-impl-idmap.c#L209-L218

I'm not all that familiar with OpenWrt with respect to it's priority inheritance implementation or lack there-of.  Depending on your use case you could just ignore the error and deal w/ possible priority inversion?  There isn't a cFS related setting to ignore it, so you'd need to locally patch.

**Reply by mc-bane** [reply]
Thanks for the response! I think that shouldn't be a problem so I patched it out of my file and was able to successfully run it on my host machine, however I've encountered a new blocker for running on the yun (pic related) where I am getting an error saying my OS_QueueCreate function isn't implemented. Looking at https://github.com/nasa/osal/blob/ce11e7a20d4fd9176469659acb1ee3b371ba4291/src/os/posix/src/os-impl-queues.c#L120-L123 it seems my mq_open function is returning an error code. Does that have something to do with the "CFE_PSP: mmap to EEPROM File failed" message I am also getting?

![Screenshot from 2021-08-13 10-41-50](https://user-images.githubusercontent.com/86273861/129374606-03d31443-00a3-4b47-aed9-513ec448e8e1.png)

**Reply by skliper** [reply]
Those are likely unrelated errors, although likely both due to functionality not supported by OpenWrt.  If there's a way to add in mqueue support to OpenWrt that's probably the easiest approach, otherwise you may need to provide a custom message queue implementation.  Note there's a related open issue - https://github.com/nasa/osal/issues/73 for user-space queues.

**Reply by mc-bane** [reply]
Ahh thanks, I will definitely investigate that route.
